File: src/main/java/io/austinzhu/type/adt/Maybe.java
package io.austinzhu.type.adt;

import io.austinzhu.type.Function;
import io.austinzhu.type.Type;
import io.austinzhu.type.functor.Functor;

public sealed interface Maybe<A extends Type<A>>
        extends Sum<Unit, A>, Type<Maybe<A>>, Functor<Maybe<A>, A>
        permits Maybe.Nothing, Maybe.Just {

    Maybe<?> NOTHING = new Nothing<>();

    static <A extends Type<A>> Maybe<A> just(A a) {
        return new Just<>(a);
    }

    final class Nothing<A extends Type<A>> extends Sum.Inl<Unit, A> implements Maybe<A> {

        private Nothing() {
            super(new Unit());
        }
    }

    final class Just<A extends Type<A>> extends Sum.Inr<Unit, A> implements Maybe<A> {

        private Just(A x) {
            super(x);
        }
    }

    @SuppressWarnings("unchecked")
    @Override
    default <B extends Type<B>, FB extends Functor<FB, B> & Type<FB>> FB fmap(Function<A, B> f, Maybe<A> fa) {
        if (fa instanceof Nothing) {
            return (FB) new Nothing<B>();
        }
        if (fa instanceof Just<A> j) {
            return (FB) new Just<>(f.apply(j.b));
        }
        throw new ClassCastException();
    }
}

File: src/main/scala/dev/austinzhu/pattern/adt/Maybe.scala
package dev.austinzhu.pattern.adt

object Maybe {

  val NOTHING: Maybe[_] = Maybe.Nothing()

  final case class Nothing[A] private() extends Maybe[A]

  final case class Just[A] private(x: A) extends Maybe[A]
}

trait Maybe[A] extends Sum[Unit, A] {

  def fromJust(x: Maybe[A]): A = x match {
    case Maybe.Just(x) => x
  }

  def fromMaybe(a: A, x: Maybe[A]): A = x match {
    case Maybe.Just(x) => x
    case Maybe.Nothing() => a
  }

  def isJust(x: Maybe[_]): Bool = x match {
    case _: Maybe.Just[_] => Bool.TRUE
    case _: Maybe.Nothing[_] => Bool.FALSE
  }

  def isNothing(x: Maybe[_]): Bool = x match {
    case _: Maybe.Just[_] => Bool.FALSE
    case _: Maybe.Nothing[_] => Bool.TRUE
  }

  def fmap[B](fn: A => B, f: Maybe[A]): Maybe[B] = f match {
    case Maybe.Just(x) => Maybe.Just(fn(x))
    case Maybe.Nothing() => Maybe.Nothing[B]()
  }
}